### ISSUE TITLE
Create "graceful shutdown" mechanism for activity workers

### ISSUE THREAD

**Issue by ryanwalls** [sclasen/swfsm#121]
Would be nice to have a method on the activity worker that gracefully shuts it down.  i.e.  It stops the pollers and waits for any work in progress to finish.  

Currently we can shutdown polling gracefully by calling the `ShutdownManager.StopPollers` for the polling side.  Didn't see anything for gracefully stopping work.  Perhaps we create another behavior in `ActivityTaskDispatcher `?

**Comment by sclasen** [comment]
@ryanwalls if you shutdown the Poller, your activity wont receive any further tasks, so I assume this about signaling the actual activity handler(s) that it should shut down, or just making sure any in-flight activities being worked complete before the `ShutdownManager.StopPollers`  call returns?

Its not particularly clear of course, but you can register arbitrary pairs of channels with the ShutdownManager.  the call to StopPollers wont return till everything registered with the poller has sent on its stopAck channel.

So should be 'possible' in userspace, if you want to give that a shot.  Your intuition around the Dispatcher is a good direction, in that you can create your own dispatcher, that registers with the shutdown manager, counts in-flight tasks, and only acks the stop once all in-flight tasks are done.

Something like this? Thoughts?

```go
type MyDispatcher struct {
   stop chan bool
   stopAck chan bool
   inFlight int64
}

m = &MyDispatcher{}
p = poller.ShutdownManager{}
p.Register("my-dispatcher", m.stop, m.stopAck)
m.Start()

func  (m *MyDispatcher)DispatchTask(t *swf.PollForActivityTaskOutput, f func(*swf.PollForActivityTaskOutput)){
    //run tasks in a new goroutine
    go func(){
          atomic.AddInt64(m.inFlight, 1)
          f() 
          atomic.AddInt64(m.inFlight,  -1)
    }
}

func (m *MyDispatcher)Start() {
   <-m.stop
   for atomic.LoadInt64(&m.inFlight) > 0 {
        time.Sleep(1 * time.Second)
   }  
  m.stopAck <- true
}
```

**Comment by ryanwalls** [comment]
@sclasen, I like that approach.  And would address this issue.  Would be a good dispatcher to add to the library.  That said, just realized what I asked for may not work in my case, haha.   I'm implementing blue/green deployments for tasks that could run for a week or more.  

My steps for a blue/green deploy (automated with ansible):
*  Call `StopPollers` on any deployed (blue) workers (I've created an HTTP endpoint on each worker for doing this)
*  When all pollers have stopped successfully, deploy new green worker(s).
*  When each blue worker finishes its inflight work, perhaps days later, return from the main function (which will make my docker containers exit.)

So I really need two stages.  `StopPollers` and then when `PollUntilShutdownBy` returns, drop into another loop (or wait on channel) waiting on inflight tasks.  

To accomplish this will probably use what you wrote as inspiration, but instead of tying into the `ShutdownManager` will have the dispatcher keep track of inflight tasks, so I can ask it if any work is inflight.  Will post back when I come up with a solution.  Feel free to close this issue, or keep it open if you do want to create a nice "graceful shutdown" method at some point.

**Comment by sclasen** [comment]
@ryanwalls you could probably just use 2 shutdown managers then?

```go
poller := &poller.ShutdownManager{}
activity := &poller.ShutdownManager{}
d := &MyDispatcher{}
asm.Register("dispatcher", d.stop, d.stopAck)
d.Start()


poller.StopPollers()
doDeployNewWorkers()
activity.StopPollers()  //Not well named for this use case I admit
os.Exit(0)

```

**Comment by ryanwalls** [comment]
Oooh, I like it. :+1: 

Thanks!

**Comment by sclasen** [comment]
closed via #123
